power_grid_model_c/power_grid_model/include/power_grid_model/main_model_wrapper.hpp

```diff
@@ -0,0 +1,136 @@
+        add_component<CompType>(std::span<typename CompType::InputType const>{components});
+    }
+    template <class CompType> void add_component(std::span<typename CompType::InputType const> components) {
+        impl().add_component<CompType>(components);
+    }
+
+    template <cache_type_c CacheType> void update_component(ConstDataset const& update_data) {
+        impl().update_component<CacheType>(update_data);
+    }
+
+    template <typename CompType, typename MathOutputType, typename OutputType>
+    void output_result(MathOutputType const& math_output, std::vector<OutputType>& target) const {
+        output_result<CompType>(math_output, std::span<OutputType>{target});
+    }
+    template <typename CompType, typename MathOutputType, typename OutputType>
+    void output_result(MathOutputType const& math_output, std::span<OutputType> target) const {
+        impl().output_result<CompType>(math_output, target.begin());
+    }
+
+    template <symmetry_tag sym>
+    MathOutput<std::vector<SolverOutput<sym>>> calculate_power_flow(Options const& options) {
+        return impl().calculate_power_flow<sym>(options);
+    }
+    template <symmetry_tag sym> void calculate_power_flow(Options const& options, MutableDataset const& result_data) {
+        return impl().calculate_power_flow<sym>(options, result_data);
+    }
+    template <symmetry_tag sym>
+    BatchParameter calculate_power_flow(Options const& options, MutableDataset const& result_data,
+                                        ConstDataset const& update_data) {
+        return impl().calculate_power_flow<sym>(options, result_data, update_data);
+    }
+    template <symmetry_tag sym>
+    MathOutput<std::vector<SolverOutput<sym>>> calculate_state_estimation(Options const& options) {
+        return impl().calculate_state_estimation<sym>(options);
+    }
+    template <symmetry_tag sym>
+    BatchParameter calculate_state_estimation(Options const& options, MutableDataset const& result_data,
+                                              ConstDataset const& update_data) {
+
+        return impl().calculate_state_estimation<sym>(options, result_data, update_data);
+    }
+    template <symmetry_tag sym>
+    MathOutput<std::vector<ShortCircuitSolverOutput<sym>>> calculate_short_circuit(Options const& options) {
+
+        return impl().calculate_short_circuit<sym>(options);
+    }
+    void calculate_short_circuit(Options const& options, MutableDataset const& result_data) {
+        return impl().calculate_short_circuit(options, result_data);
+    }
+    BatchParameter calculate_short_circuit(Options const& options, MutableDataset const& result_data,
+                                           ConstDataset const& update_data) {
+        return impl().calculate_short_circuit(options, result_data, update_data);
+    }
+
+    CalculationInfo calculation_info() const { return impl().calculation_info(); }
+
+  private:
```

Comment:
I personally like this better, but it just comes to preference I guess.

Reply:
we have a convention in the project to postpone all member variable declarations to the bottom of the class so that the publicly exposed functions come first. Especially in the pimpl pattern, that makes even more sense. In addition, in other classes, the list of members may be quite long (and yes, that is something that we want to improve but we're not there yet), so it may require the user to scroll down significantly before they end up at the public functions if we place the member variables at the top.